### PR TITLE
Freemabd1/rw 241 created api call for quick search

### PR DESCRIPTION
**Pull request by freemabd1**
This quick search api call will allow users of the cohort builder to potentially query a specific criteria tree without having to click thru each node in that tree. Only results are returned that are marked as selectable.

### CODE REVIEW

File: api/src/main/java/org/pmiops/workbench/cdr/dao/CriteriaDao.java [hunk 1]
```diff
@@ -9,7 +9,13 @@
 
 public interface CriteriaDao extends CrudRepository<Criteria, Long> {
 
-    @Query("select c from Criteria c where c.type like :type% and c.parentId = :parentId order by c.code asc")
-    List<Criteria> findCriteriaByTypeAndParentId(@Param("type") String type, @Param("parentId") Long parentId);
+    List<Criteria> findCriteriaByTypeAndParentIdOrderByCodeAsc(@Param("type") String type, @Param("parentId") Long parentId);
+
+    @Query("select c from Criteria c " +
```
Comment (danqrodney) [inline api/src/main/java/org/pmiops/workbench/cdr/dao/CriteriaDao.java]: I think we're going to want to use fulltext indexes here:

https://dev.mysql.com/doc/refman/5.6/en/innodb-fulltext-index.html

In some testing I did a while back, querying with LIKE on big tables was slow, but querying with fulltext indexes was super-fast.

This is important generally, and especially important if we're going to support auto-complete as you type (which I think will be very important for a feature like this to be usable.)

File: api/src/main/java/org/pmiops/workbench/cdr/dao/CriteriaDao.java [hunk 2]
```diff
@@ -9,7 +9,13 @@
...
+            "where c.type like :type " +
+            "and (lower(c.name) like lower(:name) or lower(c.code) like lower(:code)) " +
```
Comment (danqrodney) [inline api/src/main/java/org/pmiops/workbench/cdr/dao/CriteriaDao.java]: Not sure if it matters for performance, but I believe LIKE searches in MySQL are case-insensitive by default:

https://dev.mysql.com/doc/refman/5.6/en/case-sensitivity.html

So if you are going to use LIKE instead of MATCH (see above), you can probably drop the lower() everywhere in here.

File: api/src/main/java/org/pmiops/workbench/cdr/dao/CriteriaDao.java [hunk 3]
```diff
@@ -9,7 +9,13 @@
...
+            "and c.selectable = 1 " +
+            "order by c.code asc")
+    List<Criteria> findCriteriaByTypeAndNameOrCode(@Param("type") String type, @Param("name") String name, @Param("code") String code);
```
Comment (danqrodney) [inline api/src/main/java/org/pmiops/workbench/cdr/dao/CriteriaDao.java]: Why are name and code separate parameters here? Isn't there going to be just a single text box? Isn't querying by just name (what's actually visible) what we want anyway?

Comment (freemabd1) [reply api/src/main/java/org/pmiops/workbench/cdr/dao/CriteriaDao.java]: Corrent. Oversight by me. I was using the query by method name in which you have to use 2 different variables but have switched to using the @Query annotation. So i no longer need 2 variables.

**Comment by danqrodney** [comment]
I will refer you to:

https://pmi-engteam.slack.com/archives/C5TB0DP63/p1507298129000649

The query in question there ran against a table with 1M concept rows in 6 seconds using LIKE in Cloud SQL, and 0.03 seconds with a fulltext index. We want the latter. :)

File: api/src/main/java/org/pmiops/workbench/cdr/dao/CriteriaDao.java (hunk 2 above)
Comment (freemabd1) [reply api/src/main/java/org/pmiops/workbench/cdr/dao/CriteriaDao.java]: my test cases fail if i remove the lower function, but i do like the idea about the fulltext indexing.

**Comment by freemabd1** [comment]
Agreed! I will add fulltext indexes.

**Comment by danqrodney** [comment]
Note: you may need to do a native SQL query to make this work with Spring Data:

https://stackoverflow.com/questions/43989817/match-against-in-java-spring-data-jpa

File: api/db-cdr/changelog-schema/db.changelog-1.xml [hunk 4]
```diff
@@ -42,13 +42,10 @@
...
-    <createIndex
-            indexName="idx_type_parent_id"
-            tableName="criteria"
-            unique="false">
-      <column name="type"/>
-      <column name="parent_id"/>
-    </createIndex>
+
+    <modifySql dbms="mysql">
+      <append value="ENGINE=INNODB DEFAULT CHARSET=UTF8"/>
```
Comment (danqrodney) [inline api/db-cdr/changelog-schema/db.changelog-1.xml]: Why is this needed? I thought our DB setup script did this for us....

File: api/src/main/java/org/pmiops/workbench/cdr/dao/CriteriaDao.java [hunk 5]
```diff
@@ -9,7 +9,14 @@
...
-    List<Criteria> findCriteriaByTypeAndParentId(@Param("type") String type, @Param("parentId") Long parentId);
+    List<Criteria> findCriteriaByTypeAndParentIdOrderByCodeAsc(@Param("type") String type, @Param("parentId") Long parentId);
+
+    /** TODO: implement dynamic switching of schemas **/
+    @Query(value = "select * from cdr.criteria c " +
+            "where c.type = :type " +
+            "and match(c.name) against(:value in boolean mode) " +
+            "and c.is_selectable = 1 " +
+            "order by c.code asc", nativeQuery = true)
+    List<Criteria> findCriteriaByTypeAndNameOrCode(@Param("type") String type, @Param("value") String value);
```
Comment (danqrodney) [inline api/src/main/java/org/pmiops/workbench/cdr/dao/CriteriaDao.java]: If this only matches against name now, change the method name to findCriteriaByTypeAndName?

File: api/src/main/java/org/pmiops/workbench/api/CohortBuilderController.java [hunk 6]
```diff
@@ -140,7 +140,12 @@
 
     @Override
     public ResponseEntity<CriteriaListResponse> getCriteriaTreeQuickSearch(String type, String value) {
+        String nameOrCode = value + "*";
```
Comment (danqrodney) [inline api/src/main/java/org/pmiops/workbench/api/CohortBuilderController.java]: What if somebody enters in the value string... does that need to be escaped somehow?

File: api/src/bigquerytest/java/org/pmiops/workbench/api/CohortBuilderControllerTest.java [hunk 7]
```diff
@@ -233,7 +235,38 @@ public void getCriteriaByTypeAndParentId_Phecodes() throws Exception {
...
+                new org.pmiops.workbench.cdr.model.Criteria()
+                        .id(1L)
+                        .type("PHECODE")
+                        .name("Intestinal infection")
+                        .group(true)
+                        .selectable(true)
+                        .count("0")
+                        .conceptId("0");
+
+        when(mockCriteriaDao
```
Comment (danqrodney) [inline api/src/bigquerytest/java/org/pmiops/workbench/api/CohortBuilderControllerTest.java]: We should be able to have tests using a real criteria DAO instead of a mock... that would test that we set up our DAO logic appropriately... does that not work?

File: api/db-cdr/changelog-schema/db.changelog-1.xml (hunk 4 above)
Comment (freemabd1) [reply api/db-cdr/changelog-schema/db.changelog-1.xml]: i guess i'm not following. What db setup script does this?

File: api/src/bigquerytest/java/org/pmiops/workbench/api/CohortBuilderControllerTest.java (hunk 7 above)
Comment (freemabd1) [reply api/src/bigquerytest/java/org/pmiops/workbench/api/CohortBuilderControllerTest.java]: I agree, but since we have switched to using mysql full text search, i'm not sure how to get H2 in memory database to handle the match() and against() functions within a unit test. I definitely would like a test case for this but am punting now until i can find a solution for this.

File: api/src/main/java/org/pmiops/workbench/api/CohortBuilderController.java (hunk 6 above)
Comment (freemabd1) [reply api/src/main/java/org/pmiops/workbench/api/CohortBuilderController.java]: I need to get together with John and see what we want he functionality to be. We may have some validation in the client that will handle this? Keep in mind that this will morph over time as this is our first pass at this functionality in the cohort builder

File: api/db-cdr/changelog-schema/db.changelog-1.xml (hunk 4 above)
Comment (danqrodney) [reply api/db-cdr/changelog-schema/db.changelog-1.xml]: UTF8 is set here:

https://github.com/all-of-us/workbench/blob/master/api/db-cdr/create_db.sql#L1

InnoDB is the default engine for MySQL so I'm not sure you need to specify that. 

Did things break if you didn't include this?

File: api/src/bigquerytest/java/org/pmiops/workbench/api/CohortBuilderControllerTest.java (hunk 7 above)
Comment (danqrodney) [reply api/src/bigquerytest/java/org/pmiops/workbench/api/CohortBuilderControllerTest.java]: Boo...

http://h2-database.66688.n3.nabble.com/MySQL-Compatibility-Mode-Fulltext-MATCH-function-td4033152.html

There is an FT_SEARCH thing in H2, but it requires its own setup and has completely different SQL. It sucks that this isn't handled in a more standard fashion...

There aren't a lot of great options here. Only things I can come up with:

1) Have this test run against MySQL in docker running locally, instead of H2.
2) Generate the DAO interface using some macro which uses one set of SQL for MySQL (when running in source normally) and a different set of SQL for H2 (when running in tests.)

I don't love either of these options. I don't love just having mocks here, either. 

If we are going to use a mock for testing interactions with this method, let's make sure we can still use an in-memory DAO for testing interactions with other methods.

File: api/db-cdr/changelog-schema/db.changelog-1.xml (hunk 4 above)
Comment (freemabd1) [reply api/db-cdr/changelog-schema/db.changelog-1.xml]: I'm not sure... I was just following an example for full text search. I can remove it and see what happens.

File: api/src/bigquerytest/java/org/pmiops/workbench/api/CohortBuilderControllerTest.java (hunk 7 above)
Comment (freemabd1) [reply api/src/bigquerytest/java/org/pmiops/workbench/api/CohortBuilderControllerTest.java]: Absolutely. This is the only dao method that isn't tested against H2 currently. So if we have more of these cases pop up I think we can do something with option 1 above. I feel like option 2 only test sql that would never run in real mysql env, so seems pointless.

File: api/src/main/java/org/pmiops/workbench/cdr/dao/CriteriaDao.java (hunk 5 above)
Comment (freemabd1) [reply api/src/main/java/org/pmiops/workbench/cdr/dao/CriteriaDao.java]: I will be adding matching for code in another PR. Thanks for catching this.

File: api/db-cdr/changelog-schema/db.changelog-1.xml (hunk 4 above)
Comment (freemabd1) [reply api/db-cdr/changelog-schema/db.changelog-1.xml]: Still works after removing this modifySql section.

File: api/src/main/java/org/pmiops/workbench/api/CohortBuilderController.java (hunk 6 above)
Comment (danqrodney) [reply api/src/main/java/org/pmiops/workbench/api/CohortBuilderController.java]: OK. From here:

https://stackoverflow.com/questions/26507087/escape-string-for-use-in-mysql-fulltext-search

it sounds like basically we should remove (or not allow) any non-word characters.
